src/extensions/agent_modules/cloud_communication/adps2_mqtt_client/src/adps2_mqtt_client_module.c

```diff
@@ -76,20 +90,106 @@ ADUC_AGENT_CONTRACT_INFO g_adpsMqttClientContractInfo = {
+int ADPS_MQTT_Client_Module_Initialize(ADUC_AGENT_MODULE_HANDLE handle, void* moduleInitData);
+int ADPS_MQTT_Client_Module_Deinitialize(ADUC_AGENT_MODULE_HANDLE handle);
+int ADPS_MQTT_Client_Module_DoWork(ADUC_AGENT_MODULE_HANDLE handle);
+int ADPS_MQTT_Client_Module_GetData(
+    ADUC_AGENT_MODULE_HANDLE handle, ADUC_MODULE_DATA_TYPE dataType, int dataCount, void* data, int* dataCopied);
+int ADPS_MQTT_Client_Module_SetData(
+    ADUC_AGENT_MODULE_HANDLE handle, ADUC_MODULE_DATA_TYPE dataType, int key, void* data, int size);
+
+static ADPS_MQTT_CLIENT_MODULE_STATE* ModuleStateFromModuleHandle(ADUC_AGENT_MODULE_HANDLE handle)
+{
+    if ((handle != NULL) && (((ADUC_AGENT_MODULE_INTERFACE*)handle)->moduleData != NULL))
+    {
+        return (ADPS_MQTT_CLIENT_MODULE_STATE*)(((ADUC_AGENT_MODULE_INTERFACE*)handle)->moduleData);
+    }
+    return NULL;
+}
+
+/**
+ * @brief Create a Device Update Agent Module for IoT Hub PnP Client.
+ *
+ * @return ADUC_AGENT_MODULE_HANDLE The handle to the module.
+ */
+ADUC_AGENT_MODULE_HANDLE ADPS_MQTT_Client_Module_Create()
+{
+    ADUC_AGENT_MODULE_HANDLE handle = NULL;
+
+    ADPS_MQTT_CLIENT_MODULE_STATE* moduleState = malloc(sizeof(ADPS_MQTT_CLIENT_MODULE_STATE));
+    if (moduleState == NULL)
+    {
+        Log_Error("Failed to allocate memory for module state");
+        goto done;
+    }
+    memset(moduleState, 0, sizeof(ADPS_MQTT_CLIENT_MODULE_STATE));
+
+    ADUC_AGENT_MODULE_INTERFACE* moduleInterface = malloc(sizeof(ADUC_AGENT_MODULE_INTERFACE));
+    if (moduleInterface == NULL)
+    {
+        Log_Error("Failed to allocate memory for module interface");
+        goto done;
+    }
+    memset(moduleInterface, 0, sizeof(ADUC_AGENT_MODULE_INTERFACE));
+    moduleInterface->moduleData = moduleState;
+    moduleInterface->create = ADPS_MQTT_Client_Module_Create;
+    moduleInterface->destroy = ADPS_MQTT_Client_Module_Destroy;
+    moduleInterface->getContractInfo = ADPS_MQTT_Client_Module_GetContractInfo;
+    moduleInterface->doWork = ADPS_MQTT_Client_Module_DoWork;
+    moduleInterface->initializeModule = ADPS_MQTT_Client_Module_Initialize;
+    moduleInterface->deinitializeModule = ADPS_MQTT_Client_Module_Deinitialize;
+    moduleInterface->getData = ADPS_MQTT_Client_Module_GetData;
+    moduleInterface->setData = ADPS_MQTT_Client_Module_SetData;
+
+    handle = moduleInterface;
+
+done:
+    if (handle == NULL)
+    {
+        free(moduleState);
+        free(moduleInterface);
+    }
+    return handle;
+}
+
+/*
+ * @brief Destroy the Device Update Agent Module for IoT Hub PnP Client.
+ * @param handle The handle to the module. This is the same handle that was returned by the Create function.
+ */
+void ADPS_MQTT_Client_Module_Destroy(ADUC_AGENT_MODULE_HANDLE handle)
+{
+    Log_Info("DPS mqtt model destroy");
+    ADUC_AGENT_MODULE_INTERFACE* moduleInterface = (ADUC_AGENT_MODULE_INTERFACE*)handle;
+    if (moduleInterface != NULL)
+    {
+        free(moduleInterface->moduleData);
```

Comment:
Is it ok for the moduleData to be tied to the lifetime of the interface object? It seems better to free it earlier when it's no longer needed, or moduleData is shared state data that is tight dependency between each interface call? In either case, seems like it should be separated from interface, or good reason (other than convenience) not to?